File: src/main/java/com/board/growtime/service/UserService.java
package com.board.growtime.service;

import com.board.growtime.entity.User;
import com.board.growtime.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class UserService {

    private final UserRepository userRepository;

    /**
     * GitHub 사용자 정보로 사용자를 저장하거나 업데이트
     */
    public User saveOrUpdateUser(String githubId, String login, String name, String email,
                                String avatarUrl, String htmlUrl, String company, 
                                String location, String bio, String accessToken) {
        
        Optional<User> existingUser = userRepository.findByGithubId(githubId);
        
        if (existingUser.isPresent()) {
            // 기존 사용자 정보 업데이트
            User user = existingUser.get();
            user.updateUserInfo(name, email, avatarUrl, htmlUrl, company, location, bio);
            user.updateAccessToken(accessToken);
            
            log.info("기존 사용자 정보 업데이트: {}", login);
            return userRepository.save(user);
        } else {
            // 새 사용자 생성
            User newUser = new User(githubId, login, name, email, avatarUrl, htmlUrl, company, location, bio);
            newUser.updateAccessToken(accessToken);
            
            log.info("새 사용자 생성: {}", login);
            return userRepository.save(newUser);
        }
    }

    /**
     * GitHub ID로 사용자 찾기
     */
    @Transactional(readOnly = true)
    public Optional<User> findByGithubId(String githubId) {
        return userRepository.findByGithubId(githubId);
    }

    /**
     * GitHub 로그인으로 사용자 찾기
     */
    @Transactional(readOnly = true)
    public Optional<User> findByLogin(String login) {
        return userRepository.findByLogin(login);
    }

    /**
     * 사용자 ID로 사용자 찾기
     */
    @Transactional(readOnly = true)
    public Optional<User> findById(Long id) {
        return userRepository.findById(id);
    }

    /**
     * 사용자 존재 여부 확인
     */
    @Transactional(readOnly = true)
    public boolean existsByGithubId(String githubId) {
        return userRepository.existsByGithubId(githubId);
    }

    /**
     * 사용자 삭제
     */
    public void deleteUser(Long id) {
        userRepository.deleteById(id);
        log.info("사용자 삭제: ID {}", id);
    }
} 

File: src/main/kotlin/com/board/growtime/user/UserService.kt
package com.board.growtime.user

import com.board.growtime.common.exception.UserNotFoundException
import com.board.growtime.common.exception.InvalidUserDataException
import com.board.growtime.common.util.ValidationUtils
import com.board.growtime.user.dto.UserInfo
import com.board.growtime.user.dto.toUserInfo
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 사용자 기본 정보 관리를 담당하는 서비스
 * 
 * 학습 포인트:
 * 1. Service Layer 패턴 - 비즈니스 로직을 컨트롤러에서 분리
 * 2. 일관된 서비스 구조 - 모든 서비스가 동일한 패턴을 따름
 * 3. 예외 처리 통일 - 커스텀 예외를 사용한 명확한 에러 처리
 * 4. 검증 로직 분리 - 유틸리티 클래스를 활용한 입력값 검증
 */
@Service
class UserService(
    private val userRepository: UserRepository
) {
    private val log = LoggerFactory.getLogger(UserService::class.java)

    /**
     * 사용자 정보 조회
     */
    @Transactional(readOnly = true)
    fun getUserInfo(githubId: String): UserInfo {
        validateGitHubId(githubId)
        
        val user = userRepository.findByGithubId(githubId.trim())
            ?: run {
                log.warn("사용자 조회 실패: {}", githubId)
                throw UserNotFoundException(githubId)
            }
        
        log.debug("사용자 조회 성공: {}", githubId)
        return user.toUserInfo()
    }

    /**
     * GitHub ID 검증
     */
    private fun validateGitHubId(githubId: String) {
        if (!ValidationUtils.isValidGitHubId(githubId)) {
            throw InvalidUserDataException("올바른 GitHub ID 형식이 아닙니다: $githubId")
        }
    }
}
